### PR TITLE
Improved maintenance page handling of 503 errors [release_2.0]

### LINKED ISSUE #2196: Differentiate HTTP 503 Service Unavailable errors from Maintenance Mode
User Story: As a user, I want to receive accurate and informative error messages when something goes wrong, so that I (or an admin) can easily diagnose and correct the problem.

Open OnDemand uses Apache RewriteRules to serve the maintenance page when [Maintenance Mode](https://osc.github.io/ood-documentation/develop/customizations.html?#maintenance-mode) is enabled. It also returns an HTTP 503 (Service Unavailable) status code when serving this page.

However, the way it does so is by setting an ErrorDocument which is scoped to the VirtualHost. This has the undesirable side-effect of masking all other actual 503 errors, confusing the end user by returning the maintenance page when the service is not in fact in maintenance mode, as seen [here](https://discourse.openondemand.org/t/maintenance-mode-shown-when-you-restart-web-server-from-help-tab-rhel-8-3/1473) and [here](https://discourse.openondemand.org/t/maintenance-page-without-maintenance-enabled/1087/4).

We encountered this recently when a user had a stale `/var/run/ondemand-nginx/${USER}/passenger.sock` file; they were directed to the maintenance page due to the internal 503 error, even though maintenance mode was not enabled. It took quite a bit of troubleshooting to discover that this was in fact unrelated to Maintenance Mode (and that the socket file itself needed to be removed).

It would be preferable for Open OnDemand to return a different ErrorDocument for internal HTTP 503 errors than it does when in Maintenance Mode.

This could be done, for example, by scoping the ErrorDocument directive to the Directory rather than to the VirtualHost; then (when Maintenance Mode is enabled) redirecting with a 302 to a path inside that Directory, which then redirects to the maintenance page with a 503.

```
  RewriteRule ^.*$ /public/maintenance/ [R=302,L]
  <Directory "/var/www/ood/public/maintenance">
    RewriteCond %{REQUEST_URI} !/public/maintenance/index\.html$
    RewriteRule ^.*$ /public/maintenance/index.html [R=503,L]
    ErrorDocument 503 /public/maintenance/index.html
  </Directory>
```

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201735133575781/1202709181699823) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by treydock**
Fixes #2196

### CODE REVIEW

File: ood-portal-generator/templates/ood-portal.conf.erb [hunk 1]
```diff
@@ -86,13 +86,11 @@ Listen <%= addr_port %>
...
   RewriteCond <%= @public_root %>/maintenance/index.html -f
   RewriteCond /etc/ood/maintenance.enable -f
-  RewriteCond %{REQUEST_URI} !/public/maintenance/.*$
+  RewriteCond %{REQUEST_URI} !<%= @public_uri %>/maintenance/.*$
   <%- @maintenance_ip_whitelist.each do |ip| -%>
   RewriteCond %{REMOTE_ADDR} !^<%= escape_ip(ip) %>
   <%- end -%>
-  RewriteRule ^.*$ <%= @public_uri %>/maintenance/index.html [R=503,L]
-  ErrorDocument 503 <%= @public_uri %>/maintenance/index.html
-  Header Set Cache-Control "max-age=0, no-store"
```
Comment (johrstrom) [inline ood-portal-generator/templates/ood-portal.conf.erb:95]: We need to see how this behaves in 2.0. 2.1 already had this removed - but 2.0 didn't.

I keep deferring this test, but basically we want to be sure that the maintenance page isn't being cached forever in 2.1.  By removing the `Header Set Cache-Control` are we running into a similar situation?

Comment (treydock) [reply ood-portal-generator/templates/ood-portal.conf.erb:95]: I just tested on webdev02. I removed the whitelist for VPN so I didn't have to disconnect.  But I loaded OnDemand, touched maintenance.enable, reloaded (not full reload, just click refresh button) and got maintenance page then removed the enable and reloaded again and got dashboard.

So one thing I think this change might break is the URL for maintenance page goes from staying like `/pun/sys/dashboard` to being forced to `/public/maintenance`.  So if a user refreshes, they won't go back to dashboard, they must modify URL.  I think one way we could ease that issue is update maintenance page with very clear link to get back to OnDemand.  We'd also need to update OSC's specific page as we override OOD default page.

Comment (treydock) [reply ood-portal-generator/templates/ood-portal.conf.erb:95]: I might have a solution in mind to fix the URL issue.  I am building new nightlies for dev and will test.

Comment (treydock) [reply ood-portal-generator/templates/ood-portal.conf.erb:95]: So tested on dev with nightly I just built for Gitlab, enable maintenance and get this:

https://ondemand-dev.osc.edu/public/maintenance/index.html

Added a few lines to maintenance directory

```
    RewriteCond /etc/ood/maintenance.enable !-f
    ReWriteRule ^.*$ /
```

Removed maintenance file and reloaded, back to OnDemand.  I'll open 2.1 pull request and update this pull request.

**Review by johrstrom — APPROVED** [review]
Thanks for testing manually.

**Pull request merged**